src/session_impl.cpp

```diff
@@ -1685,6 +1691,13 @@ namespace aux {
 		this->abort();
 	}
 
+	void session_impl::on_ip_change(error_code const& ec)
+	{
+		if (ec) return;
+		m_ip_notifier.async_wait(std::bind(&session_impl::on_ip_change, this, _1));
```

Comment:
would you mind using a lambda and call ``on_ip_change`` via ``session_impl::wrap()`` here, and in the other place you call ``async_wait``? It just wraps the callback to handle exceptions, preventing them from leaking into the ``io_service``.